File: examples/csharp/Recipes/GeneratePickTicketPdf.cs
// Recipe: Generate Pick Ticket and PO PDFs
// Docs:   docs/recipes/generate-pick-ticket-pdf.md
//
// Generate a production-order pick ticket as a base64-encoded PDF via the
// dedicated report endpoint POST /api/v2/process/pdfreport (service
// m_picktickets). NOTE: m_picktickets is a WRITE — it creates the
// pick-ticket record at location_id AND returns its PDF in one call.
//
// Key rules (all verified live — see the recipe page):
//   - Wrong-endpoint trap: POST /api/v2/transaction accepts an m_* payload
//     and returns Succeeded but emits NOTHING. Reports go to
//     /api/v2/process/pdfreport.
//   - m_picktickets REQUIRES UseCodeValues: true and the code "P"
//     (Production Order); UseCodeValues: false returns HTTP 500.
//   - Status and Type are numeric 0 with Keys: [] — not the "New" shape.
//   - Prerequisite: prod_order_hdr.printed = 'Y' (run a ProductionOrder
//     transaction with print_form = ON first).
//   - Success is a JSON ARRAY; errors use the P21 error envelope
//     (ErrorType/ErrorMessage), not Summary/Messages.
//
// The pdfreport endpoint is not wrapped by P21Client, so this uses a raw
// HttpClient authenticated through the Common project's P21Config/P21Auth
// helpers (RecipeHelpers.CreateRawClientAsync).

using System.Text;
using Newtonsoft.Json.Linq;
using static P21Examples.Recipes.RecipeHelpers;

namespace P21Examples.Recipes;

public static class GeneratePickTicketPdf
{
    private const string ProdOrder = "1000123";  // production order number
    private const string LocationId = "10";      // location whose inventory the components pick from

    public static async Task RunAsync()
    {
        Console.WriteLine("Recipe - Generate Pick Ticket PDF (m_picktickets)");
        Console.WriteLine(new string('=', 60));

        var payload = new JObject
        {
            ["Name"] = "m_picktickets",
            ["UseCodeValues"] = true,   // required here -- false returns HTTP 500
            ["Transactions"] = new JArray
            {
                new JObject
                {
                    ["Status"] = 0,     // numeric 0 for report payloads
                    ["DataElements"] = new JArray
                    {
                        new JObject
                        {
                            ["Keys"] = new JArray(),  // always empty for reports
                            ["Type"] = 0,             // numeric 0 for report payloads
                            ["Name"] = "TABPAGE_1.tp_1_dw_1",
                            ["Rows"] = new JArray
                            {
                                new JObject
                                {
                                    ["Edits"] = new JArray
                                    {
                                        // code "P" = Production Order (label is rejected)
                                        Edit("create_pick_ticket_type", "P"),
                                        Edit("beg_prod_order", ProdOrder),
                                        Edit("end_prod_order", ProdOrder),
                                        Edit("location_id", LocationId),
                                    }
                                }
                            }
                        }
                    }
                }
            }
        };

        PrintPayload("m_picktickets payload (POST /api/v2/process/pdfreport)", payload);
        Console.WriteLine(
            "\nNOTE: this WRITES — it creates the pick-ticket record at the location" +
            " in addition to returning the PDF.");

        if (!ConfirmExecute())
            return;

        var (rawHttp, uiServer, _) = await CreateRawClientAsync();
        using var http = rawHttp;

        // ------------------------------------------------------------------
        // Write: generate the report (creates the ticket + returns the PDF)
        // ------------------------------------------------------------------
        var response = await http.PostAsync(
            $"{uiServer}/api/v2/process/pdfreport",  // NOT /api/v2/transaction
            new StringContent(payload.ToString(), Encoding.UTF8, "application/json"));
        var bodyText = await response.Content.ReadAsStringAsync();

        // Errors come back as the standard P21 error envelope (ErrorType/ErrorMessage),
        // NOT the Summary/Messages format used by /transaction.
        if (!response.IsSuccessStatusCode)
        {
            Console.WriteLine($"HTTP {(int)response.StatusCode}: {bodyText}");
            return;
        }

        var parsed = JToken.Parse(bodyText);
        if (parsed is JObject envelope && envelope["ErrorMessage"] != null)
        {
            Console.WriteLine($"{envelope["ErrorType"]}: {envelope["ErrorMessage"]}");
            return;
        }

        // Success is a JSON ARRAY -- even for a single document
        if (parsed is not JArray documents || documents.Count == 0)
        {
            Console.WriteLine($"No documents returned: {parsed}");
            return;
        }

        // ------------------------------------------------------------------
        // Save + verify each document
        // ------------------------------------------------------------------
        foreach (var doc in documents.OfType<JObject>())
        {
            var status = doc["ResponseStatus"]?["StatusCode"]?.ToString();
            var documentData = doc["DocumentData"]?.ToString();
            if (status != "Success" || string.IsNullOrEmpty(documentData))
            {
                var msg = doc["ResponseStatus"]?["Message"]?.ToString() ?? "Unknown error";
                Console.WriteLine($"Document failed: {msg}");
                continue;
            }

            var pdfBytes = Convert.FromBase64String(documentData);
            // FileName includes .pdf, e.g. "PPT<nnn> PRODUCTION_PICK_TICKET.pdf"
            var filename = doc["FileName"]?.ToString() ?? "pick_ticket.pdf";
            await File.WriteAllBytesAsync(filename, pdfBytes);
            Console.WriteLine($"Saved {filename} ({pdfBytes.Length} bytes)");

            // Verify: decoded bytes start with %PDF and content type is PDF
            var isPdf = pdfBytes.Length > 4 &&
                        pdfBytes[0] == (byte)'%' && pdfBytes[1] == (byte)'P' &&
                        pdfBytes[2] == (byte)'D' && pdfBytes[3] == (byte)'F';
            Console.WriteLine($"  Starts with %PDF: {(isPdf ? "yes" : "NO")}");
            Console.WriteLine($"  DocumentContentType: {doc["DocumentContentType"]}");
            Console.WriteLine($"  ResponseStatus.Message: {doc["ResponseStatus"]?["Message"]}");
        }

        Console.WriteLine(
            "\nTo prove the pick-ticket row landed, reprint it: run" +
            " m_reprintpicktickets with beg_prod_pick_ticket_no/end_prod_pick_ticket_no" +
            " set to the number from the FileName — a second PDF confirms the record.");
    }
}

File: examples/python/recipes/generate_pick_ticket_pdf.py
"""
Generate a Pick Ticket PDF (m_picktickets via /api/v2/process/pdfreport)

Generate a production-order pick ticket as a base64-encoded PDF via the
dedicated report endpoint. m_picktickets CREATES the pick-ticket record at
location_id AND returns its PDF in one call -- which is why this script is
gated behind --execute. Never post an m_* report to /api/v2/transaction: it
returns Succeeded and emits nothing.

Prerequisite: the production order's form must already be printed
(prod_order_hdr.printed = 'Y') -- run a ProductionOrder transaction with
print_form = ON first.

Mirrors: docs/recipes/generate-pick-ticket-pdf.md

Usage:
    python examples/python/recipes/generate_pick_ticket_pdf.py            # dry run (default)
    python examples/python/recipes/generate_pick_ticket_pdf.py --execute  # run report, save PDF
"""

import argparse
import base64
import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

import httpx
from common.auth import get_token, get_auth_headers, get_ui_server_url
from common.config import load_config

import warnings
warnings.filterwarnings("ignore")

# --- Configuration (generic placeholders -- substitute your own) ------------
PROD_ORDER = "1000123"  # production order number
LOCATION_ID = "10"      # location whose inventory the components pick from


def build_report_payload() -> dict:
    """Build the m_picktickets report payload.

    Reports use numeric Status/Type 0 and Keys: [] -- not the "New"
    record-edit shape. m_picktickets requires UseCodeValues: true with the
    code "P" (Production Order); False returns HTTP 500.

    Returns:
        dict: Complete pdfreport payload.
    """
    return {
        "Name": "m_picktickets",
        "UseCodeValues": True,   # required here -- False returns HTTP 500
        "Transactions": [{
            "Status": 0,         # numeric 0 for report payloads
            "DataElements": [{
                "Keys": [],      # always empty for reports
                "Type": 0,       # numeric 0 for report payloads
                "Name": "TABPAGE_1.tp_1_dw_1",
                "Rows": [{"Edits": [
                    # code "P" = Production Order (the display label is rejected)
                    {"Name": "create_pick_ticket_type", "Value": "P"},
                    {"Name": "beg_prod_order", "Value": PROD_ORDER},
                    {"Name": "end_prod_order", "Value": PROD_ORDER},
                    {"Name": "location_id", "Value": LOCATION_ID},
                ]}],
            }],
        }],
    }


def run_report(ui_server: str, headers: dict, verify_ssl: bool, payload: dict) -> list[dict]:
    """POST the report payload and return the document array.

    Errors come back as the standard P21 error envelope
    (ErrorType/ErrorMessage), NOT the Summary/Messages format of /transaction.

    Args:
        ui_server: UI server URL.
        headers: Auth headers.
        verify_ssl: Whether to verify SSL certificates.
        payload: Report payload from build_report_payload().

    Returns:
        list[dict]: Documents (success is a JSON array, even for one document).

    Raises:
        SystemExit: On HTTP errors, error envelopes, or an empty result.
    """
    response = httpx.post(
        f"{ui_server}/api/v2/process/pdfreport",  # NOT /api/v2/transaction
        headers=headers, json=payload, verify=verify_ssl,
        follow_redirects=True, timeout=120,
    )
    if response.status_code >= 400:
        raise SystemExit(f"HTTP {response.status_code}: {response.text}")
    result = response.json()
    if isinstance(result, dict) and "ErrorMessage" in result:
        raise SystemExit(f"{result.get('ErrorType')}: {result['ErrorMessage']}")
    if not (isinstance(result, list) and result):
        raise SystemExit(f"No documents returned: {result}")
    return result


def save_documents(documents: list[dict]) -> None:
    """Decode and save each returned document, verifying it is a PDF.

    Args:
        documents: Document array from run_report().
    """
    for doc in documents:
        status = doc.get("ResponseStatus", {}).get("StatusCode")
        if status != "Success" or not doc.get("DocumentData"):
            msg = doc.get("ResponseStatus", {}).get("Message", "Unknown error")
            print(f"Document failed: {msg}")
            continue
        pdf_bytes = base64.b64decode(doc["DocumentData"])
        # FileName includes .pdf, e.g. "PPT<nnn> PRODUCTION_PICK_TICKET.pdf"
        filename = doc.get("FileName", "pick_ticket.pdf")
        with open(filename, "wb") as f:
            f.write(pdf_bytes)
        # Verify: decoded bytes must start with %PDF
        is_pdf = pdf_bytes.startswith(b"%PDF")
        print(f"Saved {filename} ({len(pdf_bytes)} bytes) "
              f"{'-- verified %PDF header' if is_pdf else '-- WARNING: not a PDF'}")


def read_ticket_status(ui_server: str, headers: dict, verify_ssl: bool,
                       ticket_no: str) -> None:
    """Read the new ticket back via /transaction/get and print its status.

    Args:
        ui_server: UI server URL.
        headers: Auth headers.
        verify_ssl: Whether to verify SSL certificates.
        ticket_no: Production pick-ticket number (from the PDF FileName).
    """
    get_payload = {
        "ServiceName": "ProductionOrderPicking",
        "TransactionStates": [{
            "DataElementName": "TP_PRODPICKTICKETCONF.tp_prodpickticketconf",
            "Keys": [{"Name": "prod_pick_ticket_number", "Value": ticket_no}],
        }],
    }
    resp = httpx.post(f"{ui_server}/api/v2/transaction/get",
                      headers=headers, json=get_payload, verify=verify_ssl,
                      follow_redirects=True, timeout=60)
    resp.raise_for_status()

    for txn in resp.json().get("Transactions", []):
        for de in txn.get("DataElements", []):
            for row in de.get("Rows", []):
                fields = {e["Name"]: e["Value"] for e in row.get("Edits", [])}
                if "row_status_flag" in fields:
                    # 702 = Open, 1962 = Confirmed, 1268 = Completed
                    print(f"Ticket {fields.get('prod_pick_ticket_number')} "
                          f"for prod order {fields.get('prod_order_number')}: "
                          f"status {fields.get('row_status_flag')}")


def main() -> None:
    """Entry point: build payload; on --execute run the report and save the PDF."""
    parser = argparse.ArgumentParser(
        description="Generate a pick-ticket PDF (docs/recipes/generate-pick-ticket-pdf.md)")
    parser.add_argument("--execute", action="store_true",
                        help="Actually run the report -- this CREATES the pick-ticket "
                             "record in P21 (default: dry run)")
    args = parser.parse_args()

    print("Recipe - Generate Pick Ticket PDF (m_picktickets)")
    print("=" * 60)

    payload = build_report_payload()

    if not args.execute:
        print("\nDRY RUN - payload that would be POSTed to "
              "{ui_server}/api/v2/process/pdfreport")
        print("(side effect at execute time: the pick-ticket record is CREATED at "
              f"location {LOCATION_ID}):")
        print(json.dumps(payload, indent=2))
        print("\nRe-run with --execute to run the report.")
        return

    config = load_config()
    token_data = get_token(config)
    headers = get_auth_headers(token_data["AccessToken"])
    ui_server = get_ui_server_url(config.base_url, token_data["AccessToken"], config.verify_ssl)
    print(f"UI Server: {ui_server}")

    documents = run_report(ui_server, headers, config.verify_ssl, payload)
    save_documents(documents)

    # --- Read the new ticket back (number comes from the FileName) ---
    file_name = documents[0].get("FileName", "")
    match = re.match(r"PPT(\d+)", file_name)
    if not match:
        raise SystemExit(f"Could not extract ticket number from '{file_name}'")
    read_ticket_status(ui_server, headers, config.verify_ssl, match.group(1))


if __name__ == "__main__":
    main()
